sklearn/decomposition/dict_learning.py

```diff
@@ -309,9 +314,109 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
+    nonzero_examples = list(
+        np.where(samples_norms_squared > _ATOM_NORM_TOLERANCE)[0])
+
+    if len(nonzero_examples) < n_components:
+        # this is not effective, but still allowed
+        replace = True
+    else:
+        replace = False
+    chosen_examples = choice(nonzero_examples, replace=replace,
+                             size=n_components, random_state=random_state)
+    return X[chosen_examples, :]
+
+
+def _init_dict(X, n_components, code_init, dict_init, method, random_state):
+    """Initialize dictionary for learning."""
+
+    if code_init is not None and dict_init is not None:
+        code = np.array(code_init, order='F')
+        # Don't copy V, it will happen below
+        dictionary = dict_init
+    else:
+        if method == 'svd':
+            code, S, dictionary = linalg.svd(X, full_matrices=False)
+            dictionary = S[:, np.newaxis] * dictionary
+        elif method == 'sample':
+            n_samples = X.shape[0]
+            code = np.zeros((n_samples, n_components))
+            dictionary = _init_dict_from_samples(X, n_components, random_state)
+        else:
+            raise ValueError("Unknown init method " + method)
+
+    dict_size = len(dictionary)
+    if n_components <= dict_size:  # True even if n_components=None
+        code = code[:, :n_components]
+        dictionary = dictionary[:n_components, :]
+    else:
+        code = np.c_[code, np.zeros((len(code), n_components - dict_size))]
+        dictionary = np.r_[dictionary,
+                           np.zeros((n_components - dict_size,
+                                     dictionary.shape[1]))]
+    # Fortran-order dict, as we are going to access its row vectors
+    dictionary = np.array(dictionary, order='F')
+
+    return code, dictionary
+
+
+def _worst_represented_example(X, dictionary, sparse_codes):
+    """Find the sample that has the most representation error."""
+
+    residuals = X - np.dot(sparse_codes, dictionary)
+    errors_squared = np.sum(residuals * residuals, axis=1)
+    worst_index = np.argmax(errors_squared)
+    return X[worst_index, :]
+
+
+def _decompose_projections(residual, g_old):
+    """Decompose the residual using SVD-approximating projections."""
+
+    d = np.dot(residual, g_old)
+    d_norm = norm(d)
+    if d_norm >= _ATOM_NORM_TOLERANCE:
+        d /= d_norm
+    g = np.dot(residual.T, d)
+    return d, g
+
+
+def _ksvd_atom_update(X, atom_index, dictionary, sparse_codes):
+    """Update single dictionary atom and the corresponding codes.
+
+    Use SVD decomposition to preserve sparsity.
+
+    """
+
+    atom_usages = np.nonzero(sparse_codes[:, atom_index])[0]
+    if len(atom_usages) == 0:
+        # replace with the new atom
+        atom = _worst_represented_example(X, dictionary, sparse_codes)
+        dictionary[atom_index, :] = atom / norm(atom)
+    elif len(atom_usages) == 1:
+        # update is trivial if the atom is used in only one sample
+        atom = X[atom_usages[0], :]
```

Comment:
Not sure this is correct. The new atom is definitely not replaced by the signal using it.  Not in the original algorithm anyway.

I mean, I can see why you would reason that way: it is the only signal using this atom, so why not make the atom be exactly the signal which would lead to a perfect approximation (to equality).

The thing is that we have no guarantee that the atom at this iteration will also be used by a single signal at future iterations. That is to say, if now only one signal uses it, perhaps future changes will push others to also use it. Forcing it to be replaced with the actual signal prohibits that possibility.

Think of it this way, if the current atom is very close to another atom (their mutual coherence is almost 1), then signals will migrate from one to the other during dictionary learning iterations until they grow further apart or until one of them is replaced by the `_worst_represented_example`.